[sirui] iP

beginner-web-developer opened the pull request:
### Alice Chatbot

> “A computer is like a violin. You can imagine it making beautiful music, but you have to learn how to play it.” -- Bill Gates ([source](https://goglobalways.com/blog/inspiring-computer-science-quotes/))

It is a CLI application, as such it is **SUPER FAST** to use! Being a CLI application, which only accepts specific comments currently, it is important to know which commands to input to get the right response. Current commands include:
- "list"
- "bye"
- "todo"
- "deadline"
- "event"
- "mark"
- "unmark"
- "delete"

Features:
- [x] Managing tasks
- [ ] Managing deadlines (_coming soon_)
- [ ] Reminders (_coming soon_)

All you need to do is,
1. Download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

The code uses OOP. Here is an example snapshot of the `Alice` class using other classes such as `Scanner`, `Storage` and `Ui`.
```java
public class Alice {
    private final TaskList TASKS;
    private final Storage STORAGE;
    private final Ui UI;
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Update Alice.java to greet the user and exit

Commit:
Update Alice.java to echo the user's command and exit upon command "bye"

Commit:
Fix Alice.java to not echo the "bye" command

Commit:
Update Alice.java to be able to save tasks and list them using the "list" command

Commit:
Add ability to mark tasks as done and undone

Commit:
Add different types of tasks (event, deadline, todo) that can be added to the tasklist

Commit:
Add test scripts and update Alice.java to have scanner checks to prevent exceptions from happening

Commit:
Add error handling to Alice.java to support missing arguments and unknown commands

Commit:
Add support for error handling during marking and unmarking tasks

Commit:
Add support for deleting tasks

Commit:
Add exception handling for string arguments for index in mark, unmark and delete commands

Commit:
Refactor code to use enums for the commands

Commit:
Add ability to store and fetch tasks from previous sessions

Commit:
Update .gitignore to ignore changes to the data.txt file used for storing tasks

Commit:
Merge branch 'branch-Level-7'

Commit:
Update dates to be interpreted as java.time.LocalDate instead of a string

Commit:
Merge branch 'branch-Level-8'

Commit:
Refactor code related to loading and unloading tasks from the data file into Storage class

Commit:
Remove duplicated code

Commit:
Refactor rest of the code to Parser, Tasklist and Ui class

Commit:
Merge branch 'branch-A-MoreOOP'

Commit:
Refactor code into packages

Commit:
Fix error message printing

Commit:
Merge branch 'branch-A-Packages'

Commit:
Merge branch 'add-gradle-support'

Commit:
Add JUnit tests for packages Task and Storage

Commit:
Fix main class path

Commit:
Add JavaDoc comments to document methods and classes

Commit:
Fix code style

Commit:
Fix JavaDoc comments to follow coding standards

Commit:
Add find functionality to Alice

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/exceptions/AliceException.java
#	src/main/java/tasks/Event.java

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/commands/Command.java
#	src/main/java/parser/Parser.java
#	src/main/java/tasks/Task.java
#	src/main/java/ui/Ui.java

Commit:
Add checkstyle

Commit:
Fix style for code

Commit:
Merge branch 'branch-A-CheckStyle'

jinnsuke commented on src/main/java/Alice.java:
```diff
@@ -0,0 +1,44 @@
+        // initialise UI with the loaded tasks
+        ui = new Ui(tasks);
+    }
+
```
I like how you've structured the initialization of Storage, TaskList, and Ui in the constructor—it makes the flow of the setup quite clear. I’m wondering, though, would it be beneficial to extract the task-loading logic into a separate method? It could make the constructor more concise and improve readability. Just a thought for making the code a bit cleaner!

jinnsuke commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,136 @@
+                System.out.println(result[1] + " is not a number!");
+                System.out.println("------------------------------------------");
+                break;
+            }
```
I like how you've implemented error handling for invalid task numbers with a try-catch block! It ensures that the program doesn't crash when a user inputs a non-numeric value. However, would it be more user-friendly to handle the error by providing a more detailed message or even prompting the user to re-enter a valid number? This could improve the user experience by guiding them on how to correct their input. What do you think about this approach?

jinnsuke commented on src/main/java/storage/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+        tasks.add(task);
+        return task;
+    }
+
```
The addTask method in the TaskList class handles both task creation and addition. While this approach works, it might be worth considering separating the task creation logic into a dedicated method or class. This would enhance the Single Responsibility Principle (SRP), making the code easier to maintain, especially if task creation logic becomes more complex in the future. This separation could also make the method cleaner and more focused on adding tasks to the list. What do you think about this?

jinnsuke commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,139 @@
+    public void showDivider() {
+        System.out.println(Ui.DIVIDER);
+    }
+}
```
The Ui class is well-structured and easy to read, with each method having a clear and focused responsibility. The separation of concerns is well-executed, making the code intuitive to understand and maintain. I particularly appreciate how methods like printHandleTaskMessage and showDivider are used to encapsulate specific actions, ensuring that the code remains organised and avoiding repetitive code. This design enhances the overall readability and makes it easy to extend or modify the class in the future. Great job!

jinnsuke reviewed (commented):
LGTM. Just some simple recommendations!

Commit:
Implement simple GUI structure for Alice

Commit:
Separate GUI into controller and view

Commit:
Restore previous Alice functionality in GUI

Commit:
Fix code style

Commit:
Merge branch 'branch-A-Level-10'

Commit:
Refactor MissingArgumentException to use varargs

Commit:
Merge branch 'branch-A-Varargs'

sayomaki commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,58 @@
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
+        ObservableList<Node> tmp = FXCollections.observableArrayList(this.getChildren());
```
I know this naming could be better but the javadoc does help explain what the method does!

sayomaki replied:
I know this naming could be better but the javadoc does help explain what the method does!

sayomaki commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,68 @@
+    @FXML
+    private VBox dialogContainer;
+    @FXML
+    private TextField input;
```
For this name, it may be not so obvious what this input is referring to (until you find where the input is used). A better name could be `userInput`

sayomaki commented on src/main/java/exceptions/AliceException.java:
```diff
@@ -0,0 +1,23 @@
+     */
+    @Override
+    public String toString() {
+        return "OOPS! I don't know what '" + this.message + "' means!";
```
`this` keyword may be omitted as there is no overlapping variable with the same name.

sayomaki commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,111 @@
+     * @param size number of tasks in the list.
+     * @return tasks in the list.
+     */
+    public String getListedTasks(String tasks, int size) {
```
Maybe for this method, you can follow the naming for the other methods in the class (ending with the -Message suffix), I think this would be helpful for others to understand and to being consistent!

sayomaki reviewed (commented):
Good to see that the code is mostly following good code naming practices and the methods are very concise, but with just a few little comments as added below.

Code is also well structured and easy to understand and follow!

Commit:
Add assertions

Commit:
Refactor parser to follow code standards

Commit:
Refactor storage and tasklist classes

Commit:
Revert "Refactor storage and tasklist classes"

This reverts commit 5c3d6b9b48d10713e528a5cfc8f23c7d0e281318.

Commit:
Reapply "Refactor storage and tasklist classes"

This reverts commit 386e144c81cab13735d339e20f870b7bf01e3610.

Commit:
Add "exit" command

"bye" prints the exit message to the user, and "exit" is used to close
the GUI.

"exit" is not a command recognized by Alice, causing the "OOPS" error
message to appear as the GUI closes, which is not desirable.

Let's update commands to include parsing for "exit" command.

beginner-web-developer committed:
Merge pull request #1 from beginner-web-developer/branch-A-Assertions

Add assertions

Commit:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/storage/TaskList.java

beginner-web-developer committed:
Merge pull request #2 from beginner-web-developer/branch-A-CodeQuality

Improve code quality

Commit:
Add .yml file for CI

Commit:
Fix test

Commit:
Implement ability to add priority to task

Commit:
Update chatbot to list tasks according to priority

Tasks are listed according to when they are added.

To better improve the user experience, tasks should be grouped by
priority with HIGH priority being printed first.

Let's separate the printing of tasks based on priority.

This also allows the code to be more flexible. It is easy to update the
chatbot if new priorities are added.

Commit:
Update loading and storing of tasks to include priority

Commit:
Add new command to update priority

Commit:
Fix bug when updating priority

Commit:
Merge branch 'branch-C-Priority'

Commit:
Refactor InvalidPriorityException to print in GUI

Commit:
Change Image to circle view

Commit:
Merge branch 'branch-A-BetterGui'

Commit:
Add UI screenshot

beginner-web-developer committed:
Update README.md

Commit:
Refactor parse method

Commit:
Merge branch 'master' of https://github.com/beginner-web-developer/ip

beginner-web-developer committed:
Update README.md

beginner-web-developer committed:
Update README.md

Commit:
Fix parse bug for todo

beginner-web-developer committed:
Update README.md
